reconcile/utils/state.py

```diff
@@ -415,5 +402,50 @@ def __getitem__(self, item):
         except json.decoder.JSONDecodeError:
             raise KeyError(item)
 
-    def __setitem__(self, key, value):
+    def __setitem__(self, key: str, value: Any) -> None:
         self._set(key, value)
+
+    def transaction(self, key: str, value: Any) -> "_TransactionContext":
+        """Get a context manager to set the key in the state if no exception occurs."""
+        return _TransactionContext(self, key, value)
+
+
+class _TransactionContext:
+    """A context manager to set a key in the state if no exception occurs."""
+
+    def __init__(
+        self,
+        state: State,
+        key: str,
+        value: Any,
+    ):
+        self.state = state
+        self.key = key
+        self.value = value
+
+    def __enter__(self) -> bool:
+        """Check if the key exists and store its previous value."""
+        self._previous_value = None
+        if _exists := self.state.exists(self.key):
+            self._previous_value = self.state[self.key]
+        return _exists
+
+    def __exit__(
+        self,
+        exc_type: type[BaseException] | None,
+        exc_value: BaseException | None,
+        traceback: TracebackType | None,
+    ) -> None:
+        if exc_type:
+            # if an exception occurred, we don't want to write to the state
+            return
+
+        try:
+            # the key/value might have been changed in the meantime
+            if self.state[self.key] != self._previous_value:
+                raise RuntimeError(
+                    f"State key {self.key} has been changed during the transaction."
+                )
+        except KeyError:
+            pass
+        self.state[self.key] = self.value
```

Comment:
Thank you for highlighting this. Changed in d6fdfe1

Reply:
somehow boto3 response `Metadata` only contains user defined metadata, `etag` is in `response["Etag"]` instead of `response["Metadata"]`, need to extract another method to handle this.

Reply:
Another thing to consider is if during the transaction, some other request saved the same object again, `ETag` may not change, `LastModified` may change, should the transaction fail or not.

Reply:
just checked [doc](https://docs.aws.amazon.com/AmazonS3/latest/API/API_Object.html), we shouldn't use `etag` as it's only calculated for content, not metedata.

>ETag
The entity tag is a hash of the object. The ETag reflects changes only to the contents of an object, not its metadata. The ETag may or may not be an MD5 digest of the object data. Whether or not it is depends on how the object was created and how it is encrypted as described below:

Reply:
Change in the latest commit. Comparing the `Etag` is good enough; the important thing is that the value has not changed.

Reply:
Does the order of operations matter? Like now the value is `a`

* request 1 start first to update it to `b`
* request 2 start next update it to `a`
* somehow request 2 is fast enough to finish before request 1 can update
* then request 1 will override the value to `b`
 
Is this expected?

Reply:
The main use-case is acting if the key is not set:

```python
state = init_state(...)
with state.transaction("feature.foo.bar", True) as exists:
    if not exists:
        do_what_ever_is_needed()
        ...
```
For sure, you can update the value as well, but then we need a locking mechanism, which isn't in place. I left a note about that in the last commit.

Reply:
Then what's the use case for etag comparison? If it's only used for a single process/thread, there won't by any concurrency requests...

Reply:
Actually, this is a good point. I wanted this feature as a fail-safe mechanism during new integration deployments, but without any looking, it doesn't make any sense! 🤦

Reply:
I removed the value/etag caching and comparison stuff. I can live without it for the use-case mentioned above. We can still introduce a locking mechanism in the future if needed.

Reply:
could split it out as error handling to avoid write is already met your requirements, optimistic locking requires additional design, most ORM will use a default timestamp field like `updated_at` and add it as a `where` clause during update. Some service requires a version field to be specified like DynamoDB then do conditional update. As for S3, not that easy because there is no conditional put object in one request at all.

Reply:
I still compare the actual S3 object value and the value from `transaction` to avoid unneeded overwrites.